### ISSUE TITLE
AsyncExecutor stops executing new tasks, ends in RejectedExecutionException

### ISSUE THREAD

**Issue by rockjam** [slick/slick#1892]
Several times we had this exception in our production environment. Stacktrace is down below. After the first exception of this kind, every database operation gets rejected with the same exception, and only application restart helps. This doesn't occur often, but in some circumstances, we yet have to investigate.

Stack trace: 

```
java.util.concurrent.RejectedExecutionException: Task slick.basic.BasicBackend$DatabaseDef$$anon$2@335e47ae rejected from slick.util.AsyncExecutor$$anon$2$$anon$1@11324348[Running, pool size = 20, active threads = 0, queued tasks = 1000, completed tasks = 375006]
	at java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2047)
	at java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:823)
	at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1369)
	at slick.util.AsyncExecutor$$anon$2$$anon$3.execute(AsyncExecutor.scala:161)
	at slick.basic.BasicBackend$DatabaseDef.runSynchronousDatabaseAction(BasicBackend.scala:264)
	at slick.basic.BasicBackend$DatabaseDef.runSynchronousDatabaseAction$(BasicBackend.scala:262)
	at slick.jdbc.JdbcBackend$DatabaseDef.runSynchronousDatabaseAction(JdbcBackend.scala:37)
	at slick.basic.BasicBackend$DatabaseDef.slick$basic$BasicBackend$DatabaseDef$$runInContextInline(BasicBackend.scala:241)
	at slick.basic.BasicBackend$DatabaseDef.runInContextSafe(BasicBackend.scala:147)
	at slick.basic.BasicBackend$DatabaseDef.slick$basic$BasicBackend$DatabaseDef$$runInContextInline(BasicBackend.scala:171)
	at slick.basic.BasicBackend$DatabaseDef.runInContextSafe(BasicBackend.scala:147)
	at slick.basic.BasicBackend$DatabaseDef.slick$basic$BasicBackend$DatabaseDef$$runInContextInline(BasicBackend.scala:171)
	at slick.basic.BasicBackend$DatabaseDef.runInContextSafe(BasicBackend.scala:147)
	at slick.basic.BasicBackend$DatabaseDef.runInContext(BasicBackend.scala:141)
	at slick.basic.BasicBackend$DatabaseDef.runInContext$(BasicBackend.scala:140)
	at slick.jdbc.JdbcBackend$DatabaseDef.runInContext(JdbcBackend.scala:37)
	at slick.basic.BasicBackend$DatabaseDef.runInternal(BasicBackend.scala:76)
	at slick.basic.BasicBackend$DatabaseDef.runInternal$(BasicBackend.scala:75)
	at slick.jdbc.JdbcBackend$DatabaseDef.runInternal(JdbcBackend.scala:37)
	at slick.basic.BasicBackend$DatabaseDef.run(BasicBackend.scala:73)
	at slick.basic.BasicBackend$DatabaseDef.run$(BasicBackend.scala:73)
	at slick.jdbc.JdbcBackend$DatabaseDef.run(JdbcBackend.scala:37)
	at com.company.service.services.ReadService.find(ReadService.scala:16)
	at com.company.service.resources.EntityResource$.$anonfun$apply$3(EntityResource.scala:40)
	at akka.http.scaladsl.server.directives.OnSuccessMagnet$$anon$1.$anonfun$directive$2(FutureDirectives.scala:97)
	... 131 frames truncated
```

Our setup:
* slick 3.2.2
* slick-hikaricp 3.2.2
* number of threads and connection pool size  are 20

One interesting fact is: according to hikariCP monitoring all 20 connections from connection pool are in use during this failure(internal hikariCP monitoring used)

**Comment by rockjam** [comment]
I suspect that it's not a usual task rejection under heavy load from thread pool state:
```
pool size = 20, active threads = 0, queued tasks = 1000, completed tasks = 375006
```

**Comment by hvesalai** [comment]
As you noted, the default queue max size is 1000, so the exception is normal in this case. What is not normal is `active threads = 0`, so that is what you should concentrate on. Do you have a thread dump of this situation so we could see where all the threads are (or if they are alive in the first place). Please do not include the thread dump here directly, but instead make a gist or pastebin and add the link here.

**Comment by hvesalai** [comment]
Yes, the real question is, where are the threads. If they are not active, what are they waiting for.

**Comment by renatocaval** [comment]
#1875 (fixed In v3.2.3) could lead to the impression that we are loosing the threads. 

The bug happens when using streaming calls while having DB connection failures. After a while, the `AsyncExecutor` will believe that all threads are in use and won't take any new task. Could that be the cause of this?

**Comment by guizmaii** [comment]
@rockjam Any news ?

**Comment by guizmaii** [comment]
Yes but only in my tests for now:

```
2018-05-04 08:40:05,007 WARN  slick.basic.BasicBackend.stream - Error scheduling synchronous streaming
java.util.concurrent.RejectedExecutionException: Task slick.basic.BasicBackend$DatabaseDef$$anon$3@32476d51 rejected from slick.util.AsyncExecutor$$anon$2$$anon$1@25475727[Terminated, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 150]
	at java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2063)
	at java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:830)
	at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1379)
	at slick.util.AsyncExecutor$$anon$2$$anon$3.execute(AsyncExecutor.scala:161)
	at slick.basic.BasicBackend$DatabaseDef.scheduleSynchronousStreaming(BasicBackend.scala:301)
	at slick.basic.BasicBackend$DatabaseDef.scheduleSynchronousStreaming$(BasicBackend.scala:299)
	at slick.jdbc.JdbcBackend$DatabaseDef.scheduleSynchronousStreaming(JdbcBackend.scala:37)
	at slick.basic.BasicBackend$DatabaseDef.streamSynchronousDatabaseAction(BasicBackend.scala:294)
	at slick.basic.BasicBackend$DatabaseDef.streamSynchronousDatabaseAction$(BasicBackend.scala:292)
	at slick.jdbc.JdbcBackend$DatabaseDef.streamSynchronousDatabaseAction(JdbcBackend.scala:37)
	at slick.basic.BasicBackend$DatabaseDef.slick$basic$BasicBackend$DatabaseDef$$runInContextInline(BasicBackend.scala:239)
	at slick.basic.BasicBackend$DatabaseDef.runInContextSafe(BasicBackend.scala:147)
	at slick.basic.BasicBackend$DatabaseDef.runInContext(BasicBackend.scala:141)
	at slick.basic.BasicBackend$DatabaseDef.runInContext$(BasicBackend.scala:140)
	at slick.jdbc.JdbcBackend$DatabaseDef.runInContext(JdbcBackend.scala:37)
	at slick.basic.BasicBackend$DatabaseDef$$anon$1.subscribe(BasicBackend.scala:117)
	at akka.stream.impl.fusing.ActorGraphInterpreter$BatchingActorInputBoundary.preStart(ActorGraphInterpreter.scala:117)
	at akka.stream.impl.fusing.GraphInterpreter.init(GraphInterpreter.scala:295)
	at akka.stream.impl.fusing.GraphInterpreterShell.init(ActorGraphInterpreter.scala:554)
	at akka.stream.impl.fusing.ActorGraphInterpreter.tryInit(ActorGraphInterpreter.scala:676)
	at akka.stream.impl.fusing.ActorGraphInterpreter.preStart(ActorGraphInterpreter.scala:724)
	at akka.actor.Actor.aroundPreStart(Actor.scala:528)
	at akka.actor.Actor.aroundPreStart$(Actor.scala:528)
	at akka.stream.impl.fusing.ActorGraphInterpreter.aroundPreStart(ActorGraphInterpreter.scala:667)
	at akka.actor.ActorCell.create(ActorCell.scala:654)
	at akka.actor.ActorCell.invokeAll$1(ActorCell.scala:525)
	at akka.actor.ActorCell.systemInvoke(ActorCell.scala:547)
	at akka.dispatch.Mailbox.processAllSystemMessages(Mailbox.scala:282)
	at akka.dispatch.Mailbox.run(Mailbox.scala:223)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:234)
	at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
2018-05-04 08:40:05,401 WARN  slick.basic.BasicBackend.stream - Error scheduling synchronous streaming
java.util.concurrent.RejectedExecutionException: Task slick.basic.BasicBackend$DatabaseDef$$anon$3@491ea42e rejected from slick.util.AsyncExecutor$$anon$2$$anon$1@25475727[Terminated, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 150]
	at java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2063)
	at java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:830)
	at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1379)
	at slick.util.AsyncExecutor$$anon$2$$anon$3.execute(AsyncExecutor.scala:161)
	at slick.basic.BasicBackend$DatabaseDef.scheduleSynchronousStreaming(BasicBackend.scala:301)
	at slick.basic.BasicBackend$DatabaseDef.scheduleSynchronousStreaming$(BasicBackend.scala:299)
	at slick.jdbc.JdbcBackend$DatabaseDef.scheduleSynchronousStreaming(JdbcBackend.scala:37)
	at slick.basic.BasicBackend$DatabaseDef.streamSynchronousDatabaseAction(BasicBackend.scala:294)
	at slick.basic.BasicBackend$DatabaseDef.streamSynchronousDatabaseAction$(BasicBackend.scala:292)
	at slick.jdbc.JdbcBackend$DatabaseDef.streamSynchronousDatabaseAction(JdbcBackend.scala:37)
	at slick.basic.BasicBackend$DatabaseDef.slick$basic$BasicBackend$DatabaseDef$$runInContextInline(BasicBackend.scala:239)
	at slick.basic.BasicBackend$DatabaseDef.runInContextSafe(BasicBackend.scala:147)
	at slick.basic.BasicBackend$DatabaseDef.runInContext(BasicBackend.scala:141)
	at slick.basic.BasicBackend$DatabaseDef.runInContext$(BasicBackend.scala:140)
	at slick.jdbc.JdbcBackend$DatabaseDef.runInContext(JdbcBackend.scala:37)
	at slick.basic.BasicBackend$DatabaseDef$$anon$1.subscribe(BasicBackend.scala:117)
	at akka.stream.impl.fusing.ActorGraphInterpreter$BatchingActorInputBoundary.preStart(ActorGraphInterpreter.scala:117)
	at akka.stream.impl.fusing.GraphInterpreter.init(GraphInterpreter.scala:295)
	at akka.stream.impl.fusing.GraphInterpreterShell.init(ActorGraphInterpreter.scala:554)
	at akka.stream.impl.fusing.ActorGraphInterpreter.tryInit(ActorGraphInterpreter.scala:676)
	at akka.stream.impl.fusing.ActorGraphInterpreter.preStart(ActorGraphInterpreter.scala:724)
	at akka.actor.Actor.aroundPreStart(Actor.scala:528)
	at akka.actor.Actor.aroundPreStart$(Actor.scala:528)
	at akka.stream.impl.fusing.ActorGraphInterpreter.aroundPreStart(ActorGraphInterpreter.scala:667)
	at akka.actor.ActorCell.create(ActorCell.scala:654)
	at akka.actor.ActorCell.invokeAll$1(ActorCell.scala:525)
	at akka.actor.ActorCell.systemInvoke(ActorCell.scala:547)
	at akka.dispatch.Mailbox.processAllSystemMessages(Mailbox.scala:282)
	at akka.dispatch.Mailbox.run(Mailbox.scala:223)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:234)
	at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

It is really annoying 😕

**Comment by hvesalai** [comment]
Terminated is the key here. Somebody closes the pool before your code is done.

**Comment by guizmaii** [comment]
Ok thank you @hvesalai :)

**Comment by rockjam** [comment]
We updated slick to `3.2.3` and didn't experience this issue since then. We are monitoring it, so I'll update this issue if it comes up.

**Comment by hvesalai** [comment]
Ok. Let's close this issue latest 2018-07-01 if it has not been seen by then.

**Comment by morozov1** [comment]
I experience same bug on Slick 3.2.3 in runtime. 
`Task slick.basic.BasicBackend$DatabaseDef$$anon$2@4d13c8d7 rejected from slick.util.AsyncExecutor$$anon$2$$anon$1@5500b4ca[Running, pool size = 30, active threads = 0, queued tasks = 500, completed tasks = 65366002]`

**Comment by lilac** [comment]
I experienced the same bug on slick 3.2.3 while using prisma:1.14.
```
java.util.concurrent.RejectedExecutionException: Task slick.basic.BasicBackend$DatabaseDef$$anon$2@135816a8 rejected from slick.util.AsyncExecutor$$anon$2$$anon$1@6b63cdbc[Running, pool size = 9, active threads = 0, queued tasks = 1000, completed tasks = 3111]
	at java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2063)
	at java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:830)
	at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1379)
	at slick.util.AsyncExecutor$$anon$2$$anon$3.execute(AsyncExecutor.scala:161)
	at slick.basic.BasicBackend$DatabaseDef.runSynchronousDatabaseAction(BasicBackend.scala:264)
	at slick.basic.BasicBackend$DatabaseDef.runSynchronousDatabaseAction$(BasicBackend.scala:262)
	at slick.jdbc.JdbcBackend$DatabaseDef.runSynchronousDatabaseAction(JdbcBackend.scala:37)
	at slick.basic.BasicBackend$DatabaseDef.slick$basic$BasicBackend$DatabaseDef$$runInContextInline(BasicBackend.scala:241)
	at slick.basic.BasicBackend$DatabaseDef.runInContextSafe(BasicBackend.scala:147)
	at slick.basic.BasicBackend$DatabaseDef.run$1(BasicBackend.scala:176)
	at slick.basic.BasicBackend$DatabaseDef.slick$basic$BasicBackend$DatabaseDef$$runInContextInline(BasicBackend.scala:182)
	at slick.basic.BasicBackend$DatabaseDef.runInContextSafe(BasicBackend.scala:147)
	at slick.basic.BasicBackend$DatabaseDef.slick$basic$BasicBackend$DatabaseDef$$runInContextInline(BasicBackend.scala:206)
	at slick.basic.BasicBackend$DatabaseDef.runInContextSafe(BasicBackend.scala:147)
	at slick.basic.BasicBackend$DatabaseDef.runInContext(BasicBackend.scala:141)
	at slick.basic.BasicBackend$DatabaseDef.runInContext$(BasicBackend.scala:140)
	at slick.jdbc.JdbcBackend$DatabaseDef.runInContext(JdbcBackend.scala:37)
	at slick.basic.BasicBackend$DatabaseDef.runInternal(BasicBackend.scala:76)
	at slick.basic.BasicBackend$DatabaseDef.runInternal$(BasicBackend.scala:75)
	at slick.jdbc.JdbcBackend$DatabaseDef.runInternal(JdbcBackend.scala:37)
	at slick.basic.BasicBackend$DatabaseDef.run(BasicBackend.scala:73)
	at slick.basic.BasicBackend$DatabaseDef.run$(BasicBackend.scala:73)
	at slick.jdbc.JdbcBackend$DatabaseDef.run(JdbcBackend.scala:37)
	at com.prisma.api.connector.jdbc.impl.JdbcDatabaseMutactionExecutor.execute(JdbcDatabaseMutactionExecutor.scala:33)
	at com.prisma.api.connector.jdbc.impl.JdbcDatabaseMutactionExecutor.executeTransactionally(JdbcDatabaseMutactionExecutor.scala:19)
	at com.prisma.api.mutations.ClientMutationRunner$.$anonfun$run$2(ClientMutationRunner.scala:20)
	at scala.concurrent.Future.$anonfun$flatMap$1(Future.scala:302)
	at scala.concurrent.impl.Promise.$anonfun$transformWith$1(Promise.scala:37)
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:60)
	at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:55)
	at akka.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:91)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:81)
	at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:91)
	at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:40)
	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:43)
	at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

**Comment by dope9967** [comment]
I experienced this too, started happening yesterday. Did not yet manage to replicate in test environment. Since my database is growing each day and queries are getting more heavy by a little bit too, I guess some kind of threshold was breached. What happens functionally in application at the time is that many database actions are getting executed at same time asynchronously and some scheduled tasks are starting up streamed actions. You would think, that if slick task queue is depleted, it would just reject incoming tasks, but it stops working all together. I don't know what is the implementation on slick side, but maybe something is not thread safe with it's queue counters?
```
2018-09-04 07:14:06,776 [warn] in application-akka.actor.default-dispatcher-83 from s.b.B.stream line:28 - Error scheduling synchronous streaming
java.util.concurrent.RejectedExecutionException: Task slick.basic.BasicBackend$DatabaseDef$$anon$3@390d9497 rejected from slick.util.AsyncExecutor$$anon$2$$anon$1@66a8a3bb[Terminated, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 15352]
at java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2063)
at java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:830)
at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1379)
at slick.util.AsyncExecutor$$anon$2$$anon$3.execute(AsyncExecutor.scala:161)
at slick.basic.BasicBackend$DatabaseDef.scheduleSynchronousStreaming(BasicBackend.scala:301)
at slick.basic.BasicBackend$DatabaseDef.scheduleSynchronousStreaming$(BasicBackend.scala:299)
at slick.jdbc.JdbcBackend$DatabaseDef.scheduleSynchronousStreaming(JdbcBackend.scala:37)
at slick.basic.BasicBackend$BasicStreamingActionContext.restartStreaming(BasicBackend.scala:520)
at slick.basic.BasicBackend$BasicStreamingActionContext.cancel(BasicBackend.scala:544)
at akka.stream.impl.fusing.ActorGraphInterpreter$BatchingActorInputBoundary.onInternalError(ActorGraphInterpreter.scala:189)
at akka.stream.impl.fusing.GraphInterpreterShell.$anonfun$tryAbort$1(ActorGraphInterpreter.scala:621)
at akka.stream.impl.fusing.GraphInterpreterShell.$anonfun$tryAbort$1$adapted(ActorGraphInterpreter.scala:621)
at scala.collection.immutable.List.foreach(List.scala:389)
at akka.stream.impl.fusing.GraphInterpreterShell.tryAbort(ActorGraphInterpreter.scala:621)
at akka.stream.impl.fusing.ActorGraphInterpreter.$anonfun$postStop$1(ActorGraphInterpreter.scala:794)
at akka.stream.impl.fusing.ActorGraphInterpreter.$anonfun$postStop$1$adapted(ActorGraphInterpreter.scala:794)
at scala.collection.immutable.Set$Set1.foreach(Set.scala:95)
at akka.stream.impl.fusing.ActorGraphInterpreter.postStop(ActorGraphInterpreter.scala:794)
at akka.actor.Actor.aroundPostStop(Actor.scala:536)
at akka.actor.Actor.aroundPostStop$(Actor.scala:536)
at akka.stream.impl.fusing.ActorGraphInterpreter.aroundPostStop(ActorGraphInterpreter.scala:667)
at akka.actor.dungeon.FaultHandling.finishTerminate(FaultHandling.scala:210)
at akka.actor.dungeon.FaultHandling.terminate(FaultHandling.scala:172)
at akka.actor.dungeon.FaultHandling.terminate$(FaultHandling.scala:142)
at akka.actor.ActorCell.terminate(ActorCell.scala:433)
at akka.actor.ActorCell.invokeAll$1(ActorCell.scala:531)
at akka.actor.ActorCell.systemInvoke(ActorCell.scala:547)
at akka.dispatch.Mailbox.processAllSystemMessages(Mailbox.scala:282)
at akka.dispatch.Mailbox.run(Mailbox.scala:223)
at akka.dispatch.Mailbox.exec(Mailbox.scala:234)
at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

**Comment by hvesalai** [comment]
@rockjam have you experienced this again?

**Comment by hvesalai** [comment]
Everybody experiencing this still: please take a thread dump the next time you see this. We need to know where the threads have gone.

**Comment by dope9967** [comment]
@hvesalai maybe you have a tip on the easiest way to do a thread dump in my case? My application that experiences this issue runs on Play 2.6, and it does not crash or hang - I just detect through health check endpoint that I cannot execute queries and I restart it by invoking Linux service restart from health check service. Also I could not reproduce this in test environment, so I somehow have to code this into my production code, preferably without waiting manually for application to crash again.

**Comment by hvesalai** [comment]
In linux, you need to find out the process id and then execute `kill -3 <pid>`. This will send a signal to the JVM to print the threads to stdout. You need to capture the stdout somehow to a file, etc.

**Comment by dope9967** [comment]
@hvesalai ended up using `jstack` in Linux service script. Expected the crash to happen again today, but nothing bad happened, anyways if I'll experience this issue in the future, I'll post the link to gist with thread dump here.

**Comment by mgilbertnz** [comment]
I managed to reproduce this in my dev environment today.

```
[error] application - A fatal error occurred
java.util.concurrent.RejectedExecutionException: Task slick.basic.BasicBackend$DatabaseDef$$anon$2@6d27dcb9 rejected from slick.util.AsyncExecutor$$anon$2$$anon$1@a8384aa[Running, pool size = 20, active threads = 20, queued tasks = 998, completed tasks = 72532]
        at java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2047)
        at java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:823)
        at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1369)
        at slick.util.AsyncExecutor$$anon$2$$anon$3.execute(AsyncExecutor.scala:120)
        at slick.basic.BasicBackend$DatabaseDef.runSynchronousDatabaseAction(BasicBackend.scala:233)
        at slick.basic.BasicBackend$DatabaseDef.runSynchronousDatabaseAction$(BasicBackend.scala:231)
        at slick.jdbc.JdbcBackend$DatabaseDef.runSynchronousDatabaseAction(JdbcBackend.scala:38)
        at slick.basic.BasicBackend$DatabaseDef.runInContext(BasicBackend.scala:210)
        at slick.basic.BasicBackend$DatabaseDef.runInContext$(BasicBackend.scala:142)
        at slick.jdbc.JdbcBackend$DatabaseDef.runInContext(JdbcBackend.scala:38)
```

Here is a threaddump immediately after the above exception
https://gist.github.com/mgilbertnz/9d10a04c8a018dce9a406cd88a19f80b

Hope this helps!

**Comment by hvesalai** [comment]
@mgilbertnz more investigation is needed. All the slick threads are there waiting on a condition, so it must be that all the connections are held by someone, possibly a thread that has died. Are you using transactions anywhere in your code? I suspect that all the connections are held by a transaction that is blocked by something or that the threads executing scala code in those transactions have died without returning the connections back to the pool.

So the next thing for you to do is try to debug if there are exceptions being thrown that are not caught or if there flatMaps, etc being done with an execution context that has no threads.

**Comment by morozov1** [comment]
Any news?

**Comment by hvesalai** [comment]
https://www.youtube.com/watch?v=GQ05_gDud6w

**Comment by hvesalai** [comment]
As 3.3.0 has been released, I will again schedule this to be closed unless somebody can reproduce it on that version. Please if you report repetitions of this bug, include the details on the version you used, the stack trace, the thread dump and weather you are using reactive streams or not.

**Comment by XuefengWu** [comment]
I upgrade to 3.3.0 and this issue still here.

`Caused by: java.util.concurrent.RejectedExecutionException: Task slick.basic.BasicBackend$DatabaseDef$$anon$3@cfe16b9 rejected from slick.util.AsyncExecutor$$anon$1$$anon$2@63299a2f[Running, pool size = 20, active threads = 20, queued tasks = 1000, completed tasks = 1686]
        at java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2047)
        at java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:823)
        at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1369)
        at slick.util.AsyncExecutor$$anon$1$$anon$4.execute(AsyncExecutor.scala:161)
        at slick.basic.BasicBackend$DatabaseDef.runSynchronousDatabaseAction(BasicBackend.scala:264)
        at slick.basic.BasicBackend$DatabaseDef.runSynchronousDatabaseAction$(BasicBackend.scala:262)
        at slick.jdbc.JdbcBackend$DatabaseDef.runSynchronousDatabaseAction(JdbcBackend.scala:37)
        at slick.basic.BasicBackend$DatabaseDef.slick$basic$BasicBackend$DatabaseDef$$runInContextInline(BasicBackend.scala:241)
        at slick.basic.BasicBackend$DatabaseDef.runInContextSafe(BasicBackend.scala:147)
        at slick.basic.BasicBackend$DatabaseDef.runInContext(BasicBackend.scala:141)`

**Comment by XuefengWu** [comment]
I upgraded to 3.3.0 but this issue is till here.
BUT I can not reproduce in unit test now.

`Caused by: java.util.concurrent.RejectedExecutionException: Task slick.basic.BasicBackend$DatabaseDef$$anon$3@dc5085c rejected from slick.util.AsyncExecutor$$anon$1$$anon$2@7a6e06c8[Running, pool size = 20, active threads = 20, queued tasks = 1000, completed tasks = 18458]
        at java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2047)
        at java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:823)
        at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1369)
        at slick.util.AsyncExecutor$$anon$1$$anon$4.execute(AsyncExecutor.scala:161)
        at slick.basic.BasicBackend$DatabaseDef.runSynchronousDatabaseAction(BasicBackend.scala:264)
        at slick.basic.BasicBackend$DatabaseDef.runSynchronousDatabaseAction$(BasicBackend.scala:262)
        at slick.jdbc.JdbcBackend$DatabaseDef.runSynchronousDatabaseAction(JdbcBackend.scala:37)
        at slick.basic.BasicBackend$DatabaseDef.slick$basic$BasicBackend$DatabaseDef$$runInContextInline(BasicBackend.scala:241)
        at slick.basic.BasicBackend$DatabaseDef.runInContextSafe(BasicBackend.scala:147)
        at slick.basic.BasicBackend$DatabaseDef.runInContext(BasicBackend.scala:141)`

**Comment by hvesalai** [comment]
@XuefengWu this is normal. You are just trying to execute more jobs than your queue will fit. The bug in this report is different. The difference is in the number of active threads.

**Comment by degtevO** [comment]
@hvesalai faced the same problem like @XuefengWu with slick 3.3.1:
`java.util.concurrent.RejectedExecutionException: Task slick.basic.BasicBackend$DatabaseDef$$anon$2@66ac156f rejected from slick.util.AsyncExecutor$$anon$2$$anon$1@55b48a50[Running, pool size = 10, active threads = 10, queued tasks = 100000, completed tasks = 2792834]`

queue size is 100000. Can you indicate, where is my problem?

**Comment by hvesalai** [comment]
@degtevO you have exactly the same problem as the person above and the answer is the same. 

Note that this is not a support forum. For support, turn to one of the community forums (such as gitter).

**Comment by scand1sk** [comment]
I have ran into this issue in production recently:

```
play.api.UnexpectedException: Unexpected exception[RejectedExecutionException: Task slick.basic.BasicBackend$DatabaseDef$$anon$3@1eaa079e rejected from slick.util.As
yncExecutor$$anon$1$$anon$2@22647562[Running, pool size = 20, active threads = 0, queued tasks = 1000, completed tasks = 39742]]
```

Using Play 2.8.7 and Play-Slick 5.0.0.
I have never used debugging tools with Play before, so I do not know how to start investigating the issue. I understand that there probably is a deadlock somewhere but how can I know where the deadlock happens?
There is one place in the application where I had to use `Await(xxx, 1.minute)` because a query happens in a callback that will not accept a Future as a result. However, this does not use a transaction.

**Comment by AndreyIlin** [comment]
Faced the same issue as issue author.
`Slick version: 3.3.2`
`Streaming from Postgres`

```
2021-04-14 02:21:38,902 ERROR c.n.c.b.h.DatabaseStatusHolder - Database health check failed with an exception
java.util.concurrent.RejectedExecutionException: Task slick.basic.BasicBackend$DatabaseDef$$anon$3@6136e293 rejected from slick.util.AsyncExecutor$$anon$1$$anon$2@5349d19c[Running, pool size = 5, active threads = 0, queued tasks = 1000, completed tasks = 9577]
	at java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2063)
	at java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:830)
	at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1379)
	at slick.util.AsyncExecutor$$anon$1$$anon$4.execute(AsyncExecutor.scala:161)
	at slick.basic.BasicBackend$DatabaseDef.runSynchronousDatabaseAction(BasicBackend.scala:265)
	at slick.basic.BasicBackend$DatabaseDef.runSynchronousDatabaseAction$(BasicBackend.scala:263)
	at slick.jdbc.JdbcBackend$DatabaseDef.runSynchronousDatabaseAction(JdbcBackend.scala:37)
	at slick.basic.BasicBackend$DatabaseDef.slick$basic$BasicBackend$DatabaseDef$$runInContextInline(BasicBackend.scala:242)
	at slick.basic.BasicBackend$DatabaseDef.runInContextSafe(BasicBackend.scala:148)
	at slick.basic.BasicBackend$DatabaseDef.runInContext(BasicBackend.scala:142)
	at slick.basic.BasicBackend$DatabaseDef.runInContext$(BasicBackend.scala:141)
	at slick.jdbc.JdbcBackend$DatabaseDef.runInContext(JdbcBackend.scala:37)
	at slick.basic.BasicBackend$DatabaseDef.runInternal(BasicBackend.scala:77)
	at slick.basic.BasicBackend$DatabaseDef.runInternal$(BasicBackend.scala:76)
	at slick.jdbc.JdbcBackend$DatabaseDef.runInternal(JdbcBackend.scala:37)
	at slick.basic.BasicBackend$DatabaseDef.run(BasicBackend.scala:74)
	at slick.basic.BasicBackend$DatabaseDef.run$(BasicBackend.scala:74)
	at slick.jdbc.JdbcBackend$DatabaseDef.run(JdbcBackend.scala:37)
```
Here's gist with thread dumps for slick threads:
https://gist.github.com/AndreyIlin/43e15875f9b93cd1f2665966c3dd50d9

As far as I can see all threads are waiting for the same condition object:
`- parking to wait for <61f294a7> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)`

**Comment by hvesalai** [comment]
@AndreyIlin what's your configuration? How many connections? How many threads? If you can just show the complete configuration, that would be best in determining the cause.

**Comment by AndreyIlin** [comment]
> @AndreyIlin what's your configuration? How many connections? How many threads? If you can just show the complete configuration, that would be best in determining the cause.

Here's my slick config:
```
  db {
    keepAliveConnection = true
    numThreads = 5
    driver = "org.postgresql.Driver"
    user = ${pgsql.db_user}
    password = ${pgsql.db_pass}
    url = "jdbc:postgresql://"${pgsql.db_url}/${pgsql.db_name}"?currentSchema="${pgsql.db_name}_schema
  }
```
I'm using extended `com.github.tminglei.slickpg.ExPostgresProfile` slick profile:
```
ExPostgresProfile
    with PgCirceJsonSupport
    with PgDate2Support
    with PgArraySupport
    with PgJsonSupport
    with PgEnumSupport
    with PgSearchSupport { // ... }
    ```

**Comment by hvesalai** [comment]
@AndreyIlin your case looks like a pretty normal deadlock. I think you will find that you are invoking `run` from within an action

**Comment by AndreyIlin** [comment]
I have custom wrapper around `DBIO` it prevents calling run from inside another run, but I'll re-check.

**Comment by hvesalai** [comment]
Also within stream?

**Comment by AndreyIlin** [comment]
@hvesalai 
You're probably right. 
I'm doing repository call within database stream.
Thanks a lot for pointing out.
So this is not possible and should be omitted?

**Comment by hvesalai** [comment]
When you have a stream it means you have a connection open. Thus if all connections are streaming, there is no more connections available for other calls. This is your case.

**Comment by hvesalai** [comment]
I have solved this in my own project by having a different database connection for streaming and an other one for normal queries. That way I never get a deadlock between streaming and other queries.

**Comment by AndreyIlin** [comment]
I see, thanks for a helpful tip will try it out.
